Populate the eviction_policy field for load/store properly

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91316

This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @mlazos @soumith @yanboliang @anijain2305 @chunyuan-w @desertfire

lezcano committed:
Populate the eviction_policy field for load/store properly

This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

[ghstack-poisoned]

ngimel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -863,8 +862,12 @@ def load(self, name: str, index: sympy.Expr):
         index, mask = self.indexing(index, dense_indexing=True)
+
+        # Keep the variable in cache if we are going to reuse it
+        ep_str = "first" if name in self.current_node.last_usage else "last"
```
do we need store eviction policy? When we are writing and then reading the data, we typically need to go all the way to gmem, otherwise we would just leave it as an intermediate variable

ngimel replied:
In the current triton master (MLIR) there's no `eviction_policy` arg to store, so let's hold off on this.

lezcano replied:
ag, they said it'd be BC aaaand here we are. Leaving it as a comment in case it's relevant in the future.

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

eellison commented:
what was the benchmarking ? run torchbench ? i couldnt validate perf gain for sofmtax with this but maybe i was running wrong samples

eellison commented:
run torchbench ? i tried a few ops with this and got mixed results. what was the benchmarking script ? 

try `python /scratch/eellison/work/pytorch/benchmarks/dynamo/microbenchmarks/operatorbench.py --op=aten._softmax.default --dtype=float32 --max-samples=25 --accuracy-checking=False --suite=huggingface` before/after ?

lezcano commented:
sure, let me try that

lezcano commented:
It seems to get a consistent 1% speed-up running the command in https://github.com/pytorch/pytorch/pull/91316#issuecomment-1363421509. I have to say that the benchmarks are not as stable as I would like them to, so I run them 3 times each:
```
master :
[1.0279319318750362, 1.0543131412257343, 1.1092436595025628]
[1.029289752609728, 1.0638297295929133, 1.1131844980910401]
[1.0230071700163474, 1.0672269765613274, 1.1079364778186145]
PR :
[1.0436695871330925, 1.0672269765613274, 1.1131844980910401]
[1.0323480717808713, 1.0672269765613274, 1.1182555177307507]
[1.035240162792983, 1.0583334151878585, 1.1057159259906555]
```

ngimel commented:
What was the error you encountered?

lezcano commented:
There's this check that was triggered when I run that command
https://github.com/pytorch/pytorch/blob/f62a3cabfc53569aef37d35ee34de99ec29dc5d7/aten/src/ATen/native/SoftMax.cpp#L338-L340
now that I think about it, should that be triggered at all, given that that's the CPU path? The tests should run on CUDA, so why is this path being hit even if `half_to_float=True` in some trace?

lezcano commented:
If `eviction_policy='evict_last'` seems a bit dicey, I can always set the last read to have no policy. That should be quite uncontroversial.

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented:
`evict_first` you mean? Yeah we can leave with no policy, that shouldn't matter much. 
Can you please file an issue for the failure you encountered?

lezcano commented:
Yes, I meant `evict_first`, sorry. I did see some perf differences between one and the other in isolated experiments. In sizes for which the cache hits are important, `evict_first` was more performant. In others, I saw that setting `evict_first` was a bit of a (rather tiny) pesimisation.
Leaving it with no eviction policy in the last case. In the future we can try to predict whether caching things may be useful and use that info to populate these field. Even more, for large reductions, we could figure out whether it's hopeless to try to keep the whole input in cache and we could set the policy to `no_allocate` (if we can ask triton to support this one) for `roffset >= 64k` (or something more intelligent of course).

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -f "error coming from the SEV"

ngimel commented:
@pytorchbot revert -m "regresses performance"

ngimel commented:
@pytorchbot revert -m "regresses performance" -c nosignal

ngimel commented:
This regresses performance for hf models significantly, see https://github.com/pytorch/torchdynamo/issues/681#issuecomment-1365355831. Easy repro for torchbench:
```
python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --float16 --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt) --only hf_Bert  
```
1.19x -> 1.13x

lezcano commented:
oh boi. This is truly unexpected. Let me have a look at the generated triton code, there may be a bug in my implementation.

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -822,13 +822,15 @@ def load(self, name: str, index: sympy.Expr):
+        # https://github.com/pytorch/pytorch/pull/91316#issuecomment-1364680622
+        ep = (
+            ", eviction_policy='evict_last'"
+            if name not in self.current_node.last_usage
```
IIUC this tests if there will be future kernels that re-use the buffer being loaded. If so, this seems insufficient to me because a reduction kernel such as softmax may load the same memory location multiple times, but if fused into a single kernel this test doesn't keep it in cache.

lezcano replied:
Not quite, see how this PR modifies for how that `last_usage` node is set for nodes within a `FusedSchedulerNode`. But I think the issue similar to that one. I think this PR is not setting this correctly for nodes that are not fused.

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
Ok, I'm lost on this one. I have fixed this to take mutations into account and now the generated code for hf_Bert is what I would expect. Now, I still see a 5% regression with these changes in Bert.
When I set `eviction_policy=evict_first` for those buffers that are not going to be used within the kernel, I get a regression of 5.5%.
I have no idea what's going on here. @ngimel do you know what could be happening here or possible steps forward?

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1096,14 +1102,25 @@ def load(self, name: str, index: sympy.Expr):
+        #  1) We are in a reduction loop
+        #  2) Its not its last use
+        #  3) This load will not be lifted to body the body
+        #  4) We are not broadcasting
```
Shouldn't broadcasting benefit the most from caching since the same data is read even more times?

lezcano replied:
I got the logic wrong lol

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1096,14 +1102,25 @@ def load(self, name: str, index: sympy.Expr):
+        #  3) This load will not be lifted to body the body
+        #  4) We are not broadcasting
+        if self.inside_reduction and not self.persistent_reduction:
+            names = self.mutations or {name}
```
Won't `self.mutations` refer to every buffer used in the current loop, not just the one being loaded here?

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -233,10 +240,6 @@ def allocate(self):
                 )
             )
             and config.inplace_buffers
-            and (
-                not isinstance(V.kernel, torch._inductor.codegen.triton.TritonKernel)
-                or getattr(V.kernel, "mutations", None) is not None
-            )
```
This evaluates to `True`  as a `TritonKernel`  always has a `"mutations"` arg and it defaults to an empty set.

peterbell10 replied:
`TritonTemplateKernel` seems to default `mutations` to `None`. That might be relevant?

lezcano replied:
oops, rolling back

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Populate the eviction_policy field for load/store properly"


This helps with kernels that make use of caching like mid-range softmax
which reads the data three times.

Selecting `eviction_policy=evict_first` in the last loop of the softmax
operation seems to give a 7-10% speed-up vs. selecting `evict_last` which
was the previous option. I'll put up some benchmarks soon™.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel mlazos soumith yanboliang anijain2305 chunyuan-w desertfire

[ghstack-poisoned]

lezcano commented:
@ngimel This one is at last ready for review.
I run the benchmark suit, and we're getting around a 1% speed-up on torchbench (I don't know if it's just noise though, I hope not!). See the benchmarks [here](https://hud.pytorch.org/benchmark/compilers?startTime=Sat%2C%2027%20May%202023%2008%3A42%3A55%20GMT&stopTime=Sat%2C%2003%20Jun%202023%2008%3A42%3A55%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/Lezcano/166/orig&lCommit=3579f624718ccce83116616be3fd0aec1a00de1b&rBranch=main&rCommit=c75e064dd6a2f800476bc84d4f27d7f49cedd055).

lezcano commented:
@pytorchbot merge

desertfire commented:
@lezcano , I have confirmed that this PR caused a big regression on `pyhpc_equation_of_state` inference performance: 12.8520 → 1.9600 . It will great if we can forward fix this. cc @eellison @jansel @zou3519 

https://hud.pytorch.org/benchmark/torchbench/inductor_no_cudagraphs?startTime=Fri%2C%2002%20Jun%202023%2017%3A23%3A58%20GMT&stopTime=Fri%2C%2009%20Jun%202023%2017%3A23%3A58%20GMT&granularity=hour&mode=inference&dtype=amp&lBranch=main&lCommit=07104ca99c9d297975270fb58fda786e60b49b38&rBranch=main&rCommit=87cbfe957a3473ebd3491f8096e37084c53c4f0a

zou3519 commented:
Should we attempt to revert this or do a forward fix?

desertfire commented:
> pyhpc_equation_of_state

Given `pyhpc_equation_of_state` is not a very typical model test and training perf looks fine, I think forward fix is ok.

lezcano commented:
I am going to be on PTO for the next 3 weeks, so I'd say it's best to revert and I'll then look into this when I'm back. Sorry for that, the benchmarks in https://github.com/pytorch/pytorch/pull/91316#issuecomment-1574792030 looked alright!

lezcano commented:
Note to self: Perhaps we could get some extra speed-up by using `cache.cs` for data that will just be used once.

lezcano commented:
FWIW, I believe that the performance drop of `pyhpc_equation_of_state` was fixed in https://github.com/pytorch/pytorch/pull/103514.
